From a0a4359342ee8847df911405880c3e2e2411d09e Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Mon, 24 Jul 2023 14:57:25 +0530
Subject: [PATCH] ath12k: allocate rx_stats while enabling ext_rx_stats

rx_stats structure is specific for each station. while enabling ext_rx_stats
through debugfs, this rx_stats gets filled by the value from the monitor ring.
While station associates, without checking ext_rx_stats debugfs state,
memory for rx_stats gets allocated which is unneeded.

Fix it by checking the ext_rx_stats state before allocating the memory for
rx_stats which helps to provide more headroom incase of multiclient scenario.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -7030,7 +7030,7 @@ static int ath12k_mac_station_add(struct
 		goto exit;
 	}
 
-	if (!arsta->rx_stats) {
+	if (ath12k_debugfs_is_extd_rx_stats_enabled(ar) && !arsta->rx_stats) {
 		arsta->rx_stats = kzalloc(sizeof(*arsta->rx_stats), GFP_KERNEL);
 		if (!arsta->rx_stats) {
 			ret = -ENOMEM;
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -3419,15 +3419,12 @@ static void ath12k_dp_mon_rx_update_peer
 	u32 num_msdu;
 	u32 bw_offset;
 
-	if (!rx_stats)
-		return;
-
 	arsta->last_tx_pkt_bw = ppdu_info->bw;
 	bw_offset = arsta->last_tx_pkt_bw * 3;
 	arsta->rssi_comb = ppdu_info->rssi_comb;
 	ewma_avg_rssi_add(&arsta->avg_rssi, ppdu_info->rssi_comb + bw_offset);
 
-	if (!ath12k_debugfs_is_extd_rx_stats_enabled(ar))
+	if (!ath12k_debugfs_is_extd_rx_stats_enabled(ar) || !rx_stats)
 		return;
 
 	num_msdu = ppdu_info->tcp_msdu_count + ppdu_info->tcp_ack_msdu_count +
